pkgs/stdenv/linux/default.nix

```diff
@@ -258,10 +262,14 @@ in
   # If we ever need to use a package from more than one stage back, we
   # simply re-export those packages in the middle stage(s) using the
   # overrides attribute and the inherit syntax.
-  #
-  # resulting stage1 stdenv:
-  # - coreutils, binutils, glibc, gcc: from bootstrapFiles
-  (prevStage: stageFun prevStage {
+  (prevStage:
+    # previous stage0 stdenv:
+    assert isFromBootstrapFiles prevStage.binutils.bintools;
+    assert isFromBootstrapFiles prevStage."${localSystem.libc}";
+    assert isFromBootstrapFiles prevStage.gcc-unwrapped;
+    assert isFromBootstrapFiles prevStage.coreutils;
+    assert isFromBootstrapFiles prevStage.gnugrep;
```

Comment:
I think the asserts are useful and look reasonable to be merged as is.

I wonder if something more ergonomic could be added into `stageFun` itself, like a set of attribute names to verify the asserts. The caller change could look like:

before:

```
  (prevStage:
    # previous stage0 stdenv:
    assert isFromBootstrapFiles prevStage.binutils.bintools;
    assert isFromBootstrapFiles prevStage."${localSystem.libc}";
    assert isFromBootstrapFiles prevStage.gcc-unwrapped;
    assert isFromBootstrapFiles prevStage.coreutils;
    assert isFromBootstrapFiles prevStage.gnugrep;
stageFun prevStage {
```

After

```nix
(prevStage: stageFun prevStage {
   comesFromBootstrapFiles = with prevStage; [
     binutils.bintools
     "${localSystem.libc}"
     gcc-unwrapped
     coreutils
     gnugrep
    ...
  ];
```

Don't know if `comesFromBootstrapFiles` is trivially implementable.

Reply:
I actually wrote it this way in my first draft!

Unfortunately it is *really* hard to read:

https://github.com/NixOS/nixpkgs/commit/3d33cf83b28806ab1f4130cb1ae2e2f1b9983b01#diff-bf918d7f48f6a34a4a1acb33373dee717cff2c3f5503a7ca41d90e5f289faf1e

Since these are assertions, I felt that it was important that it should be totally obvious what is being `assert`ed, because there is no "meta-`assert`" to check that the assertions are right!

I think if we had 20 or 30 stdenv stages then it would be worth the complexity.

Reply:
Ah, yeah. Negative assertions would not fit nicely either. Never mind me then.